diff --git a/target/linux/rockchip/armv8/base-files/etc/board.d/01_leds b/target/linux/rockchip/armv8/base-files/etc/board.d/01_leds
index 3505c4ca8a09c449f00293dc32b058d4e31521ce..ec28b1ce41f43ab9f955faf67eda3e444026bba0 100644
--- a/target/linux/rockchip/armv8/base-files/etc/board.d/01_leds
+++ b/target/linux/rockchip/armv8/base-files/etc/board.d/01_leds
@@ -44,1 +44,5 @@
+hinlink,h28k)
+	ucidef_set_led_netdev "wan" "WAN" "blue:wan" "eth1"
+	ucidef_set_led_netdev "lan" "LAN" "amber:lan" "eth0"
+	;;
 esac
diff --git a/target/linux/rockchip/armv8/base-files/etc/board.d/02_network b/target/linux/rockchip/armv8/base-files/etc/board.d/02_network
index 637c428adeea70fe0334298c845887e41a0bb2f8..e6b5ca5505868284a8ac9fcc0325c799d3d648b7 100644
--- a/target/linux/rockchip/armv8/base-files/etc/board.d/02_network
+++ b/target/linux/rockchip/armv8/base-files/etc/board.d/02_network
@@ -25,3 +25,4 @@ rockchip_setup_interfaces()
+	hinlink,h28k|\
 	armsom,sige3|\
 	friendlyarm,nanopi-r5c|\
 	lunzn,fastrhino-r66s|\
diff --git a/target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity b/target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity
index d69532b9dd79c319aad2e414ace9950972db4fd3..1b9231a6e296f06cf8809e3e9e50b987c2b79733 100644
--- a/target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity
+++ b/target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity
@@ -40,2 +40,3 @@
+hinlink,h28k|\
 lunzn,fastrhino-r66s|\
 mmbox,anas3035|\
