diff --git a/target/linux/rockchip/image/armv8.mk b/target/linux/rockchip/image/armv8.mk
index 13260a23e5cc80112406409ae046ee5a86193258..98631e8996f69a0a1963bc16dde5cdb75caa4989 100644
--- a/target/linux/rockchip/image/armv8.mk
+++ b/target/linux/rockchip/image/armv8.mk
@@ -17,3 +17,15 @@
+define Device/hinlink_h28k
+  $(Device/rk3528)
+  DEVICE_VENDOR := HINLINK
+  DEVICE_MODEL := H28K
+  DEVICE_DTS := rk3528-hinlink-h28k
+  DEVICE_DTS_DIR := $$(DTS_DIR)/rockchip
+  KERNEL := kernel-bin | lzma | fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb
+  UBOOT_DEVICE_NAME := hinlink-h28k-rk3528
+  DEVICE_PACKAGES := kmod-input-adc-keys kmod-saradc-rockchip kmod-r8169
+endef
+TARGET_DEVICES += hinlink_h28k
+
 define Device/ariaboard_photonicat
   DEVICE_VENDOR := Ariaboard
   DEVICE_MODEL := Photonicat
