diff --git a/target/linux/rockchip/armv8/base-files/etc/init.d/phy-leds b/target/linux/rockchip/armv8/base-files/etc/init.d/phy-leds
index 03cb8546366108274bdb78905fa5827fd2be95ae..81dd0a12a4e7ce903529a6465cd85d87e5f530a3 100755
--- a/target/linux/rockchip/armv8/base-files/etc/init.d/phy-leds
+++ b/target/linux/rockchip/armv8/base-files/etc/init.d/phy-leds
@@ -100,2 +100,8 @@
+	hinlink,h28k)
+		netled_set "enp1s0-1::lan" "" "link_10 link_100 link_1000" "1" "rx tx" "0"
+		netled_set "enp1s0-2::lan" "" "link_10 link_100 link_1000" "0" "rx tx" "1"
+		netled_set "stmmac-0:01:green:lan" "eth0" "link_10 link_100 link_1000" "1" "rx tx" "0"
+		netled_set "stmmac-0:01:amber:lan" "eth0" "link_10 link_100 link_1000" "0" "rx tx" "1"
+		;;
 	esac
 }
